detection_rules/integrations.py

```diff
@@ -0,0 +1,135 @@
+    owner = fields.Dict(required=True)
+
+    @post_load
+    def transform_policy_template(self, data, **kwargs):
+        data["policy_templates"] = [policy["name"] for policy in data["policy_templates"]]
+        return data
+
+
+def build_integrations_manifest(token: str, overwrite: bool) -> None:
+    """Builds a new local copy of manifest.yaml from integrations Github."""
+    manifest_file_path = get_etc_path('integration-manifests.json.gz')
+    if overwrite:
+        if os.path.exists(manifest_file_path):
+            os.remove(manifest_file_path)
+    rule_integrations = os.listdir(INTEGRATION_RULE_DIR)
+    if "endpoint" in rule_integrations:
+        rule_integrations.remove("endpoint")
+
+    final_integration_manifests = {key: dict() for key in rule_integrations}
+    for ints in rule_integrations:
+        integration_manifests = get_integration_packages(token=token, org="elastic", repo="package-storage",
+                                                         branch="production", folder=f"packages/{ints}")
+        for int_man in integration_manifests:
+            validated_manifest = IntegrationManifestSchema(unknown=EXCLUDE).load(int_man)
+            package_version = validated_manifest.pop("version")
+            final_integration_manifests[ints][package_version] = validated_manifest
+
+    manifest_file = gzip.open(manifest_file_path, "w+")
+    manifest_file_bytes = json.dumps(final_integration_manifests).encode("utf-8")
+    manifest_file.write(manifest_file_bytes)
+
+
+def find_least_compatible_version(package: str, integration: str,
+                                  current_stack_version: str, packages_manifest: dict) -> str:
+    """Finds least compatible version for specified integration based on stack version supplied."""
+    integration_manifests = packages_manifest[package]
+    least_compatible_version = None
+    compatible_versions = dict()
+
+    def compare_versions(int_ver: str, pkg_ver: str) -> bool:
+        """Compares integration and package version"""
+        pkg_major, pkg_minor = Version(pkg_ver)
+        int_major, int_minor = Version(int_ver)[:2]
+
+        if int(int_major) < int(pkg_major) or int(pkg_major) > int(int_major):
+            return(False)
+        compatible = Version(int_ver) <= Version(pkg_ver)
+        return(compatible)
+
+    for ver, manifest in integration_manifests.items():
+        kibana_compat_vers = re.sub(r"\>|\<|\=|\^", "", manifest["conditions"]["kibana.version"])
+        kibana_compat_vers = kibana_compat_vers.split(" || ")
+        bool_checks = list()
+        for kcv in kibana_compat_vers:
+            bool_checks.append(compare_versions(kcv, current_stack_version))
+        if any(bool_checks):
+            compatible_versions.setdefault(ver, True)
+
+    if len(list(compatible_versions.keys())) > 0:
+        compatible_version_list = list(compatible_versions.keys())
+        compatible_version_list.sort(key=Version)
```

Comment:
Tested a few rules with this and they seem to sort fine. Chose to debug with an AWS rule because it has so many package versions we would be able to see an unsorted array of semantic versions.

<img width="238" alt="Screen Shot 2022-07-18 at 1 02 41 PM" src="https://user-images.githubusercontent.com/99630311/179564650-7080f030-5258-415e-88d4-a781b278fbcb.png">

After running and stepping over the order is correctly sorted. Do you have any other suggestions for testing? Stepping into the custom Version class, it looks like a tuple is made from the semantic version string, so when sort is called it is sorting a list of tuples with integers in them so it should sort these accordingly?

<img width="251" alt="Screen Shot 2022-07-18 at 1 02 22 PM" src="https://user-images.githubusercontent.com/99630311/179564587-0b679b53-150c-46da-812a-658e80f0e5f9.png">

Reply:
oh great, disregard then 👍